extensions/pkg/controller/common/checkerwatchdog.go

```diff
@@ -0,0 +1,193 @@
+	timer.Stop()
+	go func() {
+		for {
+			// Wait for a timer event or a result request
+			var ch chan struct{}
+			select {
+			case <-ctx.Done():
+				return
+			case reset := <-w.timerChan:
+				// Stop the timer, then reset it if a reset was requested
+				// Timers are not concurrency safe, therefore all timer updates are performed here
+				timer.Stop()
+				if reset {
+					timer.Reset(w.interval)
+				}
+				continue
+			case ch = <-w.resultChan:
+				// If the last result is not older than w.interval, use it
+				if !time.Now().After(w.resultTime.Add(w.interval)) {
+					ch <- struct{}{}
+					continue
+				}
+			case <-timer.C():
+				timer.Reset(w.interval)
+			}
+
+			// Call checker and update result
+			timeoutCtx, cancel := context.WithTimeout(ctx, w.timeout)
+			result, err := w.checker.Check(timeoutCtx)
+			cancel()
+			w.setResult(result, err)
+
+			// If a result was requested, notify the requester that the new result is available
+			if ch != nil {
+				ch <- struct{}{}
+			}
+
+			// If the check failed or returned false, cancel all contexts
+			if err != nil || !result {
+				w.logger.Info("Cancelling all contexts")
+				w.ctxCancelFuncsMutex.Lock()
+				for _, cancelFunc := range w.ctxCancelFuncs {
+					cancelFunc()
+				}
+				w.ctxCancelFuncsMutex.Unlock()
+			}
+		}
+	}()
+}
+
+// Stop stops the goroutine started by Start.
+func (w *checkerWatchdog) Stop() {
+	w.logger.V(1).Info("Stopping watchdog")
+	w.cancelFunc()
+}
+
+// AddContext adds the given context to this watchdog mapped to the given key,
+// and returns a new context that will be cancelled if the condition check fails or returns false.
+// It returns true if the first context has been added, false otherwise.
+func (w *checkerWatchdog) AddContext(ctx context.Context, key string) (context.Context, bool) {
+	w.logger.V(1).Info("Adding context", "key", key)
+
+	w.ctxCancelFuncsMutex.Lock()
+	ctx, cancelFunc := context.WithCancel(ctx)
+	w.ctxCancelFuncs[key] = cancelFunc
+	firstAdded := len(w.ctxCancelFuncs) == 1
+	w.ctxCancelFuncsMutex.Unlock()
+
+	// Reset the timer if the first context has been added
+	if firstAdded {
+		w.timerChan <- true
+	}
+
+	return ctx, firstAdded
+}
+
+// RemoveContext removes the context mapped to the given key from this watchdog.
+// It returns true if the last context has been removed, false otherwise.
+func (w *checkerWatchdog) RemoveContext(key string) bool {
+	w.logger.V(1).Info("Removing context", "key", key)
+
+	w.ctxCancelFuncsMutex.Lock()
+	delete(w.ctxCancelFuncs, key)
+	lastRemoved := len(w.ctxCancelFuncs) == 0
+	w.ctxCancelFuncsMutex.Unlock()
+
+	// Stop the timer if the last context has been removed
+	if lastRemoved {
+		w.timerChan <- false
+	}
+
+	return lastRemoved
+}
+
+// Result returns the result of the last condition check.
+func (w *checkerWatchdog) Result() (bool, error) {
+	// Request the result and wait for it to be updated if needed
+	ch := make(chan struct{})
+	w.resultChan <- ch
+	<-ch
```

Comment:
As discussed, let's try something like this instead:
```suggestion
	w.resultChan <- struct{}{}
	<-w.resultReadyChan
```

Reply:
Done